In [1]:
!pip install pandas
!pip install numpy

In [1]:
### Importing the dependencies ###
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

C:\Users\RL465FV\Anaconda3\lib\site-packages\scipy\__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.26.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
#### Loading the dataset using pandas library ####

df = pd.read_csv(r'C:\Users\RL465FV\OneDrive - EY\Desktop\My_Python\Credit Card Project\creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

Variables/Columns present here mean-

Time: Here it is the time elapsed in seconds from the time elapsed from after the first transaction which is the first row/record here. For eg: If 10th transaction has 60 as time that means the 10th transcation happened 60 seconds after the first transaction (top row/record).

Columns (V1- V28) - these are the details about the transactions and credit card which were sensitive so the data provider converted these into numeric values using PCA(Principal Component Analysis)

Amount - Its in USD. 

Class - It has 0- Genuine/Legit transaction and 1- Fraudulent tranasaction. 

In [3]:
### Bottom of the dataset ###
df.tail(10)

Time         V1         V2        V3        V4        V5  \
284797  172782.0  -0.241923   0.712247  0.399806 -0.463406  0.244531   
284798  172782.0   0.219529   0.881246 -0.635891  0.960928 -0.152971   
284799  172783.0  -1.775135  -0.004235  1.189786  0.331096  1.196063   
284800  172784.0   2.039560  -0.175233 -1.196825  0.234580 -0.008713   
284801  172785.0   0.120316   0.931005 -0.546012 -0.745097  1.130314   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284797 -1.343668  0.929369 -0.206210  0.106234  ... -0.228876 -0.514376   
284798 -1.014307  0.427126  0.121340 -0.285670  ...  0.099936  0.337120   
284799  5.519980 -1.518185  2.080825  1.159498  ...  0.103302  0.654850   
284800 -0.726571  0.017050 -0.118228  0.435402  ... -0.268048 -0.717211   
284801 -0.235973  0.812722  0.115093 -0.204064  ... -0.314205 -0.808520   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284797  0.279598  0.371441 -0.559238  0.113144  0.131507  0.081265    5.49   
284798  0.251791  0.057688 -1.508368  0.144023  0.181205  0.215243   24.05   
284799 -0.348929  0.745323  0.704545 -0.127579  0.454379  0.130308   79.99   
284800  0.297930 -0.359769 -0.315610  0.201114 -0.080826 -0.075071    2.68   
284801  0.050343  0.102800 -0.435870  0.124079  0.217940  0.068803    2.69   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284797      0  
284798      0  
284799      0  
284800      0  
284801      0  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[10 rows x 31 columns]

Here the time of last transaction is 172792 seconds, which means it has trnasactnion of 47.999777 hours i.e approx 2 days.

In [4]:
### Dataset information ###
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
### Another way of Checking the number of missing values in each column ###
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
### Distribution of legit and fraudelent transactions ###
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

### The data is highly imbalanced
This data is imbalanced because more than 99% of data is present in one class (0) i.e. normal transactions. Thus, we cannot feed this data and train the machine learning model because it will not be able to recognise the fraudelent data as the data point is very low in this class (1).

In [7]:
## Separating the data based on Legit (0) and Fraudelent(1) transaction for data analysis ##
legit = df[df.Class == 0]
fraud = df[df.Class == 1]
legit.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [8]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [9]:
### Statistical Measures of the data ###
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

Here, percentile means
25% of the data is below $5.65

In [10]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [11]:
## Comparing the values for both the transactions ##
df.groupby(df.Class).mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-sampling:
Since it is highly imbalanced data and is unfit to go ahead. So here using Under-Sampling technique to correct the imbalanced data.


Now building a sample data from the original data for Normal/Legit and Fraudelent txns.

No of Fraudelent txns - 492

No of Normal/Legit txns - 284315

So, now from these 284315 normal txns 492 random txns will be taken out and then the ratio of fraudelent and normal txns will be 50:50 which will be better for training the machine learning model.

In [12]:
### Getting 492 random records from the Legit records ###
legit_sample = legit.sample(n=492)
legit_sample.shape

(492, 31)

In [13]:
## Concatenating the fraud and legit_sample tables ###

new_df = pd.concat([legit_sample, fraud], axis = 0)
new_df.shape

(984, 31)

Here, axis = 0 means rows and axis = 1 means columns. So, here the the records will get added row-wise 1 by 1 and not column-wise.

In [14]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
25161   33526.0  1.374970 -0.486942  0.280799 -0.796257 -0.681513 -0.410692   
114666  73572.0 -0.749433  0.083954  2.280585  1.168390  0.331448  0.718667   
17059   28402.0 -0.539729  0.665452  1.238467 -0.087009  0.629944 -0.541068   
143790  85624.0  1.274190  0.074730  0.158107  0.324761 -0.145791 -0.228547   
58389   48362.0 -0.333186  0.218106  1.642252 -0.693812 -0.245003  0.733598   

              V7        V8        V9  ...       V21       V22       V23  \
25161  -0.535136 -0.077746 -0.841769  ... -0.033928 -0.189510  0.063163   
114666 -0.525059  0.487138 -0.101683  ...  0.161332  0.555090 -0.025586   
17059   0.762667  0.018879 -0.688496  ... -0.005087 -0.192679  0.136834   
143790 -0.154148  0.074539  0.055603  ... -0.276198 -0.899823  0.038616   
58389   0.150784  0.238064 -0.016993  ...  0.171842  0.511535 -0.169033   

             V24       V25       V26       V27       V28  Amount  Class  
25161  -0.283237  0.290578 -0.369064  0.020146  0.015055   19.95      0  
114666 -0.277358 -0.418169 -0.306868  0.215469  0.159973    9.99      0  
17059  -0.034546 -0.665520 -0.920409  0.167999  0.192922   32.97      0  
143790 -0.540845  0.249862  0.133108 -0.043217 -0.002539    1.98      0  
58389  -0.733018 -0.032573  1.118041 -0.033177  0.005754   78.20      0  

[5 rows x 31 columns]

In [15]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

So here as legit_sample was put first into concat hence that is the one on top of this table (new_df).

In [16]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

So here we can see now our data is uniformly distributed i.e. 492 txns each for legit and fraudulent class.

In [17]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93993.837398  0.039300 -0.006952  0.039237 -0.138553 -0.011579   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.097369 -0.022152  0.007581 -0.018098  ... -0.014119  0.004848   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.025458 -0.015160 -0.009495  0.010286 -0.016202 -0.012396 -0.002332   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       77.400000  
1      122.211321  

[2 rows x 30 columns]

Here after the comparing the mean of the legit txns the values are very much similar. This means the property/characteristic of the legit txns is same which will help the machine learning model to differentiate between the legit and fraud txn.

This step is important which helps us check whther the samples pulled are right for going ahead or not.

In [18]:
### Splitting the data into features(X) and target(Y) ###
X = new_df.drop(columns = 'Class', axis = 1)
Y = new_df['Class']

In [19]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
25161    33526.0  1.374970 -0.486942  0.280799 -0.796257 -0.681513 -0.410692   
114666   73572.0 -0.749433  0.083954  2.280585  1.168390  0.331448  0.718667   
17059    28402.0 -0.539729  0.665452  1.238467 -0.087009  0.629944 -0.541068   
143790   85624.0  1.274190  0.074730  0.158107  0.324761 -0.145791 -0.228547   
58389    48362.0 -0.333186  0.218106  1.642252 -0.693812 -0.245003  0.733598   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [20]:
print(Y)

25161     0
114666    0
17059     0
143790    0
58389     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [21]:
## Splitting the data into training & testing data ##
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.2, stratify = Y , random_state= 2)

Here, the train test split is 8:2. 

sratify(Y) based on Y will evenly distribute the data, otherwise the Y - 0 & 1 distribution will be very different in the training and testing data.

random_state = 2 ==> This helps control the random shuffling of the data. This means that if the same 'random_state' value is used in multiple runs of the train-test split function, you will get the same split of data each time.

In [22]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Training the model on X_train & Y_train. Logistic Regression is used here because for binary classification its the best suitable model. Other models can be used too but accuracy will be different.

In [23]:
### Training the model ###

model = LogisticRegression() ## creating Logistic regression object

model.fit(X_train, Y_train)  ## Training the model on training data


LogisticRegression()

In [24]:
### Model Evaluation based on accuracy score ###

## Accuracy on training data

X_train_prediction = model.predict(X_train) ## the predicted values for X_train will get stored here

Accuracy_training_data = accuracy_score(X_train_prediction, Y_train) ##predicted value will get compared with Y_train(original values)

In [25]:
print('Training data accuracy:', Accuracy_training_data)

Training data accuracy: 0.951715374841169


In [26]:
## Accuracy on test data ##

X_test_prediction = model.predict(X_test) ## the predicted values for X_test will get stored here

Accuracy_test_data = accuracy_score(X_test_prediction, Y_test) ## comparison b/w Y_test and X_test_prediction (predicted values)

print('Test data accuracy:', Accuracy_test_data)

Test data accuracy: 0.9238578680203046


Here, this model is perfect and the accuracy score of training and testing data is almost same.

Overfitting - If Train_accuracy_score > Test_accuracy_score --- Where model gets overtrained on training data

Underfitting - If Train_accuracy_score < Test_accuracy_score --- Where the model gets underfit data.


# Other models below

Here, in this project our purpose is to classify if a txn is fraudelent(1) or not (0).

Machine Learning Models:

1. Linear Regression: This model is best fit where our goal is to predict a numerical value(Y/dependent variable) based on diff. independent variables(X) such as, Fish weight based on species and its dimensions, car mileage prediction based on type, engine capacity, brand, etc.

2. Logistic Regression: This model is best fit when our goal is to predict Binary Classification. Binary classification (0/1), (True/False), (Yes/No). Ex: Heart attack prediction, fraudulent transaction, etc.

3. Decision Tree: This is best fit for classification purpose. It is a graphical representation of a series of decisions and their results. Ex: Flower species prediction based on it petal length, sepal length,  petal width, sepal width, etc.

4. Random Forest: This can be used for both classification & regression purpose. It builds multiple decisions tress and combines their output for final predition. These multiple decision trees are trained on a random subset of training data. This ensures individual trees are diverse and do not overfit on the training data. Overfitting means memorising the training data and not able to learn to predict the unseen data. 
    
    *Classification- The most common output is that which receives the most votes from individual trees.
    
    *Regression- The output is the average of the predictions from the individual trees.
    
    * Advantages of Random Forest - a.) Accuracy is high as outputs of multiple trees is used. b.) It can handle outliers/noise as well because these trees are trained on random subset of training data. c.) Large Data sets work best here.
    
    
5. K-Means clustering: This is an unsupervised machine learning model which is used for clustering or grouping the data.
